File: 20-ValidParentheses/20-ValidParentheses.java
// Last updated: 8/2/2025, 4:54:18 PM
class Solution {
    public boolean isValid(String s) {
        char[] chars = s.toCharArray();
        Stack<Character> res = new Stack<>();
        Map<Character, Character> map = new HashMap<>();
        map.put(')', '(');
        map.put(']', '[');
        map.put('}', '{');
        for (Character c : chars) {
            if (map.containsKey(c)) {
                if (res.isEmpty()) {
                    return false;
                }
                if (res.peek().equals(map.get(c))) {
                    res.pop();
                } else {
                    return false;
                }
            } else {
                res.push(c);
            }
        }
        return res.size() == 0;
    }
}

File: 20-ValidParentheses/20-ValidParentheses.py
# Last updated: 8/2/2025, 4:54:19 PM
class Solution:
    def isValid(self, s: str) -> bool:
        stack = [0]
        numOpen = 0
        numClose = 0
        for brack in s:
            if brack == '{':
                stack.append('}')
                numOpen+=1
            elif brack == '[':
                stack.append(']')
                numOpen+=1
            elif brack == '(':
                stack.append(')')
                numOpen+=1
            else:
                numClose+=1
                if brack == stack[-1]:
                    stack.pop()
                    
        return len(stack) == 1 and numOpen == numClose
        